### DISCUSSION TITLE
TFT_eSPI library and Arduino MCP2515 CAN interface library conflict on ESP32-S3 (works on ESP32) with ILI9341

### DISCUSSION THREAD

**Discussion by cvanaalst** [Bodmer/TFT_eSPI#2973] [Q&A - General]
Hi,
I need some help or guidance on how to find a solution for the following problem;
I've created a program to read CANBUS data (via Car OBD2 bus) to display it on a 2.8" TFT-screen (ILI9341) on a ESP32 Dev Kit. I am using TFT_eSPI and (https://github.com/autowp/arduino-mcp2515) each using a different SPI-interface (HSPI & VSPI). This works, but I needed some more mem for my screen (full screen sprite for anti-flicker), so I ordered a ESP32S3 Dev kit (N16R8) with more mem. I ran examples from both libraries to test if everything was working on the new Dev kit and it did.
But when I recompiled my full CANBUS-reader program, my TFT-screen just lights up and nothing displays. I do not get any error messages and my program is working correctly (debugmessages in Serial Monitor).
After debugging and creating some test programs, I've discovered that instantiation of the CANBUS object (MCP2515 mcp2515(SPI_CS) causes this behaviour. I've been trying to change the sequence of the instantiations (TFT_eSPI before or after MCP2515) as sometimes this solves the problem, => no difference.
Can anyone help me in how to find the/a solution or the exact issue/problem that causes this?

thx

Chris

**Comment by Bodmer** [comment]
Have you tried taking one of the working CAN bus examples with no changes, then slowly adding bits of TFT related code?

First just add:
#include <TFT_eSPI.h>
TFT_eSPI tft = TFT_eSPI();

**Reply by cvanaalst** [reply]
Bodmer,

I did that, I took the CAN-write.ino example tested it on both ESP32 and ESP32S3, both work. I took your TFT_Print_test.ino eample and tested it on both ESP32 & ESP32S3; both work.
With ESP32 both examples work.
With ESP32S3:
Test with CAN-write example (you can even test it without the board attached, it will print messages in the serial monitor)
I included:
- your library : CAN messages sent - screen white
- invoked it: CAN messages sent - screen white
- tft.init(); in setup(): CAN messages sent - screen white
- tft.setRotation(3); CAN messages sent - screen white
- tft.fillScreen(TFT_GREEN); CAN messages sent - BUT screen remains white

Test with TFT_Print_Test:
I included:
- mcp2515.h: correct display on TFT-screen
- invoked it: MCP2515 mcp2515(10); screen white

I use 2 different SPI-interfaces: (MISO not used for TFT)
ESP32S3 + User_Setups/Setup70b_ESP32_S3_ILI9341.h>    // Setup file for ESP32 S3 with SPI ILI9341
Default SPI settings (MCP2515): 
MOSI: 11
MISO: 13
SCK: 12
SS: 10
=========================3
TFT-eSPI SPI settings:
MOSI: 17
MISO: 17
SCK: 18
SS: 15
DC: 4
RST: 2
SPI_FREQUENCY: 40000000
SPI_READ_FREQUENCY: 6000000
SPI_TOUCH_FREQUENCY: 2500000
FLASH: 16777216
Total heap: 394452
Free heap: 369736

ESP32 + <User_Setups/Setup42_ILI9341_ESP32.h>           // Setup file for ESP32 and SPI ILI9341 240x320
Default SPI settings (MCP2515):
MOSI: 23
MISO: 19
SCK: 18
SS: 5
=========================3
TFT-eSPI SPI settings:
MOSI: 13
MISO: -1
SCK: 14
SS: 15
DC: 4
RST: 2
SPI_FREQUENCY: 27000000
SPI_READ_FREQUENCY: 20000000
SPI_TOUCH_FREQUENCY: 2500000
FLASH: 4194304
Total heap: 376108
Free heap: 350864
